File: event_hub/lib/models/event.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'ticket.dart';

class Event {
  final String eventId;
  final String organisateurId;
  final String titre;
  final String description;
  final String categorie;
  final String imageURL;
  final String lieu;
  final GeoPoint? localisation;
  final DateTime dateDebut;
  final DateTime dateFin;
  final int capaciteTotale;
  final int placesRestantes;
  final bool estPublie;
  final String statut;
  final List<String> tags;
  final double? averageRating;
  final int? reviewCount;
  final DateTime createdAt;
  final DateTime updatedAt;

  Event({
    required this.eventId,
    required this.organisateurId,
    required this.titre,
    required this.description,
    required this.categorie,
    this.imageURL = '',
    required this.lieu,
    this.localisation,
    required this.dateDebut,
    required this.dateFin,
    required this.capaciteTotale,
    required this.placesRestantes,
    this.estPublie = false,
    this.statut = 'draft',
    this.tags = const [],
    this.averageRating,
    this.reviewCount,
    required this.createdAt,
    required this.updatedAt,
  });

  bool get isPublished => estPublie && statut == 'published';
  bool get isCancelled => statut == 'cancelled';
  bool get isCompleted => statut == 'completed';
  bool get isDraft => statut == 'draft';
  bool get isAvailable => isPublished && placesRestantes > 0;
  bool get isSoldOut => placesRestantes <= 0;
  bool get isPast => dateFin.isBefore(DateTime.now());
  bool get isUpcoming => dateDebut.isAfter(DateTime.now());

  double get occupancyRate {
    if (capaciteTotale == 0) return 0;
    return ((capaciteTotale - placesRestantes) / capaciteTotale) * 100;
  }

  factory Event.fromMap(Map<String, dynamic> map) {
    return Event(
      eventId: map['eventId'] ?? '',
      organisateurId: map['organisateurId'] ?? '',
      titre: map['titre'] ?? '',
      description: map['description'] ?? '',
      categorie: map['categorie'] ?? '',
      imageURL: map['imageURL'] ?? '',
      lieu: map['lieu'] ?? '',
      localisation: map['localisation'],
      dateDebut: map['dateDebut']?.toDate() ?? DateTime.now(),
      dateFin: map['dateFin']?.toDate() ?? DateTime.now(),
      capaciteTotale: map['capaciteTotale'] ?? 0,
      placesRestantes: map['placesRestantes'] ?? 0,
      estPublie: map['estPublie'] ?? false,
      statut: map['statut'] ?? 'draft',
      tags: List<String>.from(map['tags'] ?? []),
      averageRating: map['averageRating']?.toDouble(),
      reviewCount: map['reviewCount'],
      createdAt: map['createdAt']?.toDate() ?? DateTime.now(),
      updatedAt: map['updatedAt']?.toDate() ?? DateTime.now(),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'eventId': eventId,
      'organisateurId': organisateurId,
      'titre': titre,
      'description': description,
      'categorie': categorie,
      'imageURL': imageURL,
      'lieu': lieu,
      'localisation': localisation,
      'dateDebut': dateDebut,
      'dateFin': dateFin,
      'capaciteTotale': capaciteTotale,
      'placesRestantes': placesRestantes,
      'estPublie': estPublie,
      'statut': statut,
      'tags': tags,
      'averageRating': averageRating,
      'reviewCount': reviewCount,
    };
  }

  Event copyWith({
    String? titre,
    String? description,
    String? categorie,
    String? imageURL,
    String? lieu,
    GeoPoint? localisation,
    DateTime? dateDebut,
    DateTime? dateFin,
    int? capaciteTotale,
    int? placesRestantes,
    bool? estPublie,
    String? statut,
    List<String>? tags,
    double? averageRating,
    int? reviewCount,
  }) {
    return Event(
      eventId: eventId,
      organisateurId: organisateurId,
      titre: titre ?? this.titre,
      description: description ?? this.description,
      categorie: categorie ?? this.categorie,
      imageURL: imageURL ?? this.imageURL,
      lieu: lieu ?? this.lieu,
      localisation: localisation ?? this.localisation,
      dateDebut: dateDebut ?? this.dateDebut,
      dateFin: dateFin ?? this.dateFin,
      capaciteTotale: capaciteTotale ?? this.capaciteTotale,
      placesRestantes: placesRestantes ?? this.placesRestantes,
      estPublie: estPublie ?? this.estPublie,
      statut: statut ?? this.statut,
      tags: tags ?? this.tags,
      averageRating: averageRating ?? this.averageRating,
      reviewCount: reviewCount ?? this.reviewCount,
      createdAt: createdAt,
      updatedAt: DateTime.now(),
    );
  }
}

class EventWithTickets {
  final Event event;
  final List<Ticket> tickets;

  EventWithTickets({required this.event, required this.tickets});

  factory EventWithTickets.fromMap(
    Map<String, dynamic> map,
    List<Ticket> tickets,
  ) {
    return EventWithTickets(event: Event.fromMap(map), tickets: tickets);
  }
}

File: event_hub_spring/src/main/java/com/eventhub/entity/Event.java
package com.eventhub.entity;

import jakarta.persistence.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "events")
@EntityListeners(AuditingEntityListener.class)
public class Event {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "event_id", updatable = false, nullable = false)
    private String eventId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "organisateur_id", nullable = false)
    private User organisateur;

    @Column(nullable = false)
    private String titre;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false)
    private String categorie;

    @Column(name = "image_url")
    private String imageURL;

    @Column(nullable = false)
    private String lieu;

    @Column(name = "latitude")
    private Double latitude;

    @Column(name = "longitude")
    private Double longitude;

    @Column(name = "date_debut", nullable = false)
    private LocalDateTime dateDebut;

    @Column(name = "date_fin", nullable = false)
    private LocalDateTime dateFin;

    @Column(name = "capacite_totale", nullable = false)
    private Integer capaciteTotale;

    @Column(name = "places_restantes", nullable = false)
    private Integer placesRestantes;

    @Column(name = "est_publie", nullable = false)
    private Boolean estPublie = false;

    @Column(nullable = false)
    private String statut = "draft";

    @ElementCollection
    @CollectionTable(name = "event_tags", joinColumns = @JoinColumn(name = "event_id"))
    @Column(name = "tag")
    private List<String> tags = new ArrayList<>();

    @Column(name = "average_rating")
    private Double averageRating;

    @Column(name = "review_count")
    private Integer reviewCount;

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "event", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<Ticket> tickets = new ArrayList<>();

    @OneToMany(mappedBy = "event", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<Review> reviews = new ArrayList<>();

    public Event() {}

    public String getEventId() { return eventId; }
    public void setEventId(String eventId) { this.eventId = eventId; }
    public User getOrganisateur() { return organisateur; }
    public void setOrganisateur(User organisateur) { this.organisateur = organisateur; }
    public String getTitre() { return titre; }
    public void setTitre(String titre) { this.titre = titre; }
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    public String getCategorie() { return categorie; }
    public void setCategorie(String categorie) { this.categorie = categorie; }
    public String getImageURL() { return imageURL; }
    public void setImageURL(String imageURL) { this.imageURL = imageURL; }
    public String getLieu() { return lieu; }
    public void setLieu(String lieu) { this.lieu = lieu; }
    public Double getLatitude() { return latitude; }
    public void setLatitude(Double latitude) { this.latitude = latitude; }
    public Double getLongitude() { return longitude; }
    public void setLongitude(Double longitude) { this.longitude = longitude; }
    public LocalDateTime getDateDebut() { return dateDebut; }
    public void setDateDebut(LocalDateTime dateDebut) { this.dateDebut = dateDebut; }
    public LocalDateTime getDateFin() { return dateFin; }
    public void setDateFin(LocalDateTime dateFin) { this.dateFin = dateFin; }
    public Integer getCapaciteTotale() { return capaciteTotale; }
    public void setCapaciteTotale(Integer capaciteTotale) { this.capaciteTotale = capaciteTotale; }
    public Integer getPlacesRestantes() { return placesRestantes; }
    public void setPlacesRestantes(Integer placesRestantes) { this.placesRestantes = placesRestantes; }
    public Boolean getEstPublie() { return estPublie; }
    public void setEstPublie(Boolean estPublie) { this.estPublie = estPublie; }
    public String getStatut() { return statut; }
    public void setStatut(String statut) { this.statut = statut; }
    public List<String> getTags() { return tags; }
    public void setTags(List<String> tags) { this.tags = tags; }
    public Double getAverageRating() { return averageRating; }
    public void setAverageRating(Double averageRating) { this.averageRating = averageRating; }
    public Integer getReviewCount() { return reviewCount; }
    public void setReviewCount(Integer reviewCount) { this.reviewCount = reviewCount; }
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
    public List<Ticket> getTickets() { return tickets; }
    public void setTickets(List<Ticket> tickets) { this.tickets = tickets; }
    public List<Review> getReviews() { return reviews; }
    public void setReviews(List<Review> reviews) { this.reviews = reviews; }

    public boolean isPublished() { return estPublie != null && estPublie && "published".equals(statut); }
    public boolean isCancelled() { return "cancelled".equals(statut); }
    public boolean isCompleted() { return "completed".equals(statut); }
    public boolean isDraft() { return "draft".equals(statut); }
    public boolean isAvailable() { return isPublished() && placesRestantes != null && placesRestantes > 0; }
    public boolean isSoldOut() { return placesRestantes != null && placesRestantes <= 0; }
    public boolean isPast() { return dateFin != null && dateFin.isBefore(LocalDateTime.now()); }
    public boolean isUpcoming() { return dateDebut != null && dateDebut.isAfter(LocalDateTime.now()); }
    public double getOccupancyRate() {
        if (capaciteTotale == null || capaciteTotale == 0) return 0;
        return ((capaciteTotale - placesRestantes) / (double) capaciteTotale) * 100;
    }
}